libs/core/tests/sharing_tests.rs

```diff
@@ -329,6 +329,213 @@ fn write_document_in_rejected_shared_folder_in_rejected_share_folder() {
+        .unwrap();
+    cores[1]
+        .write_document(link.id, b"document content by sharee")
+        .unwrap();
+
+    cores[1].sync(None).unwrap();
+    cores[0].sync(None).unwrap();
+
+    assert_eq!(cores[1].read_document(link.id).unwrap(), b"document content by sharee");
+    assert_eq!(cores[0].read_document(document1.id).unwrap(), b"document content by sharee");
+}
+
+#[test]
+fn write_target_by_sharee() {
+    let cores: Vec<Core> = vec![test_core_with_account(), test_core_with_account()];
+    let accounts = cores
+        .iter()
+        .map(|core| core.get_account().unwrap())
+        .collect::<Vec<_>>();
+
+    let document1 = cores[0].create_at_path("/document1").unwrap();
+
+    cores[0]
+        .share_file(document1.id, &accounts[1].username, ShareMode::Write)
+        .unwrap();
+
+    cores[0].sync(None).unwrap();
+    cores[1].sync(None).unwrap();
+
+    let link = cores[1]
+        .create_link_at_path("/link1", document1.id)
+        .unwrap();
+    cores[1]
+        .write_document(document1.id, b"document content by sharee")
+        .unwrap();
+
+    cores[1].sync(None).unwrap();
+    cores[0].sync(None).unwrap();
+
+    assert_eq!(cores[1].read_document(link.id).unwrap(), b"document content by sharee");
+    assert_eq!(cores[0].read_document(document1.id).unwrap(), b"document content by sharee");
+}
+
+#[test]
+fn create_document_in_link_folder_by_sharee() {
+    let cores: Vec<Core> = vec![test_core_with_account(), test_core_with_account()];
+    let accounts = cores
+        .iter()
+        .map(|core| core.get_account().unwrap())
+        .collect::<Vec<_>>();
+
+    let folder1 = cores[0].create_at_path("/folder/").unwrap();
+
+    cores[0]
+        .share_file(folder1.id, &accounts[1].username, ShareMode::Write)
+        .unwrap();
+
+    cores[0].sync(None).unwrap();
+    cores[1].sync(None).unwrap();
+
+    let link = cores[1].create_link_at_path("/link1", folder1.id).unwrap();
+    let result = cores[1]
+        .create_file("document1", link.id, FileType::Document)
+        .unwrap_err();
+
+    assert_eq!(result.kind, CoreError::FileNotFolder);
+}
+
+#[test]
+fn create_document_in_link_folder_by_sharer() {
+    let cores: Vec<Core> = vec![test_core_with_account(), test_core_with_account()];
+    let accounts = cores
+        .iter()
+        .map(|core| core.get_account().unwrap())
+        .collect::<Vec<_>>();
+
+    let folder1 = cores[0].create_at_path("/folder/").unwrap();
+
+    cores[0]
+        .share_file(folder1.id, &accounts[1].username, ShareMode::Write)
+        .unwrap();
+
+    cores[0].sync(None).unwrap();
+    cores[1].sync(None).unwrap();
+
+    let link = cores[1].create_link_at_path("/link1", folder1.id).unwrap();
+    cores[1].sync(None).unwrap();
+    cores[0].sync(None).unwrap();
+    let result = cores[0]
+        .create_file("document1", link.id, FileType::Document)
+        .unwrap_err();
+
+    assert_eq!(result.kind, CoreError::FileNonexistent);
```

Comment:
shouldn't this be: `FileNotFolder`?

Reply:
prob a bug

Reply:
probably we're doing something in the wrong order

Reply:
the sharer is the one who is trying to create a document under a link. The link doens't exist on their tree, it only exists for the sharee. So `FileNoneexistent` is reasonable?

Reply:
oh true fine